### ISSUE TITLE
Async Find Index

### ISSUE THREAD

**Issue by Sykander** [Sykander/patterns#5]
#4 is a dependency and must be completed before this ticket can begin
The branch associated to #4 should be merged into this branch also.

Implement the logic for Async Find Index function. The implementation should pass all tests in branch for #4 .

note: logic should be very similar to #2

**Comment by Sykander** [comment]
#16
